while-loops/while-loops.md

```diff
@@ -0,0 +1,257 @@
+* must eventually be modified in the while loop
+* typically initialized before the loop
+
+In this case, `i` is the counter variable. It gets setup before the while loop begins, and then increments inside the while loop. When we forget to change our counter variable, we often get infinite loops and bugs!
+
+Callout: An alternative syntax is to use for loops with a range. The for loop syntax sets up and modifies something like a counter variable!
+
+### `break` and `continue`
+
+`break` is a keyword. When our code comes across `break` in a while loop, it will exit the entire while loop. The next line that runs is the line after the while loop.
+
+``` python
+def demonstrate_break():
+    i = 3
+
+    while i < 10:
+        if i == 7:
+            print("i is seven! Let's get out of here!")
+            break
+        print(f"Counting: i is {i}")
+        i += 1
+
+demonstrate_break()
+```
+
+Possible example use cases:
+
+* needing to find a specific element, then leave the iteration
+* needing to leave the loop in case it comes across some invalid or error state
+
+`continue` is a keyword that will stop executing the current iteration, and resume execution at the next iteration. When our code sees `continue` , it doesn't run anymore code in that iteration, but continues to the next loop.
+
+``` python
+def demonstrate_continue():
+    i = 3
+
+    while i < 10:
+        i += 1
+        if i == 7:
+            print("i is seven!")
+            print("Let's print this special message, and then move on")
+            print("*~*~*~* \o/ *~*~*~*")
+            continue
+        print(f"Counting: i is {i}")
+
+demonstrate_continue()
+```
+
+Possible example uses cases:
+
+* wanting a loop to "skip" in a special case
+
+## Check for Understanding
+
+<!-- Need to figure out the best, most meaningful way to do a question in Learn -->
+
+<!-- question stubs -->
+
+Fix this loop so it uses the counter variable `i` , and runs while `i` is less than `5` .
+
+``` python
+def loop_until_five():
+    i = 0
+
+    while False:
+        i += 1
+
+    return i
+
+result = loop_until_five()
+```
+
+<!-- Answer: i should be 5 -->
+
+Which of these loop table options most accurately describes this loop?
+
+``` python
+i = 0
+result = None
+
+while i < 5:
+    if (i % 2) == 0:
+        result = True
+    else:
+        result = False
+    i += 1
+```
+
+<!-- Answer: This is the correct answer. False answers should toggle True/False on result, or be off-by-one on the i increments -->
+
+| Value of `i` at the beginning of the loop | `i < 5` ? | `i` at the end of loop body | `result` at end of loop body |
+|-------------------------------------------|----------|-----------------------------|------------------------------|
+| `0` | `True` | `1` | `True` |
+| `1` | `True` | `2` | `False` |
+| `2` | `True` | `3` | `True` |
+| `3` | `True` | `4` | `False` |
+| `4` | `True` | `5` | `True` |
+| `5` | `False` | Loop body doesn't execute   | -                            |
```

Comment:
```suggestion
### !challenge

* type: checkbox
* id: f4376535-5a51-4c84-9749-bfe7b9411728
* title: Loop Table

##### !question

Which of these loop table options most accurately describes this loop?

``` python
i = 0
result = None

while i < 5:
    if (i % 2) == 0:
        result = True
    else:
        result = False
    i += 1
```

##### !end-question

##### !options

* 
| Value of `i` at the beginning of the loop | `i < 5` ? | `i` at the end of loop body | `result` at end of loop body |
|-------------------------------------------|----------|-----------------------------|------------------------------|
| `1` | `True` | `1` | `True` |
| `2` | `True` | `2` | `False` |
| `3` | `True` | `3` | `True` |
| `4` | `True` | `4` | `False` |
| `5` | `False` | Loop body doesn't execute   | -                            |

* 
| Value of `i` at the beginning of the loop | `i < 5` ? | `i` at the end of loop body | `result` at end of loop body |
|-------------------------------------------|----------|-----------------------------|------------------------------|
| `1` | `True` | `1` | `True` |
| `2` | `True` | `2` | `False` |
| `3` | `True` | `3` | `True` |
| `4` | `True` | `4` | `False` |
| `5` | `True` | `5` | `True` |
| `6` | `False` | Loop body doesn't execute   | -                            |
* 
| Value of `i` at the beginning of the loop | `i < 5` ? | `i` at the end of loop body | `result` at end of loop body |
|-------------------------------------------|----------|-----------------------------|------------------------------|
| `0` | `True` | `1` | `True` |
| `1` | `True` | `2` | `False` |
| `2` | `True` | `3` | `True` |
| `3` | `True` | `4` | `False` |
| `4` | `True` | `5` | `True` |
| `5` | `False` | Loop body doesn't execute   | -                            |

##### !end-options

##### !answer

* 

| Value of `i` at the beginning of the loop | `i < 5` ? | `i` at the end of loop body | `result` at end of loop body |
|-------------------------------------------|----------|-----------------------------|------------------------------|
| `0` | `True` | `1` | `True` |
| `1` | `True` | `2` | `False` |
| `2` | `True` | `3` | `True` |
| `3` | `True` | `4` | `False` |
| `4` | `True` | `5` | `True` |
| `5` | `False` | Loop body doesn't execute   | -                            |


##### !end-answer

### !end-challenge
```

Reply:
@tildeee Learn doesn't seem to like putting tables as challenge options for me. Is that true for you too? Maybe I'm doing something wrong here.